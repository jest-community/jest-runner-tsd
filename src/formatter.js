const { posix, relative, sep } = require('path');
const { codeFrameColumns } = require('@babel/code-frame');
const ts = require('@tsd/typescript');
const chalk = require('chalk');
/** @typedef {import('tsd-lite').TsdResult} TsdResult */

const NOT_EMPTY_LINE_REGEXP = /^(?!$)/gm;
const INDENT = '  ';

/**
 * @param {string} lines
 * @param {number} level
 */
const indentEachLine = (lines, level) =>
  lines.replace(NOT_EMPTY_LINE_REGEXP, INDENT.repeat(level));

const BULLET = '\u25cf ';

/**
 * @param {string} title
 */
const makeTitle = title => indentEachLine(BULLET + title + '\n', 1);

/**
 * @param {string} input
 */
const normalizeSlashes = input => input.split(sep).join(posix.sep);

/**
 * @param {ts.SourceFile} file
 * @param {number | undefined} start
 */
function getCodeFrameAndLocation(file, start) {
  if (start === undefined) return;

  const { line, character } = file.getLineAndCharacterOfPosition(start);

  const codeFrame = codeFrameColumns(
    file.text,
    { start: { line: line + 1, column: character + 1 } },
    { highlightCode: true, linesAbove: 0, linesBelow: 0 }
  );

  const location =
    chalk.dim('at ') +
    chalk.cyan(normalizeSlashes(relative('', file.fileName))) +
    chalk.dim(':' + (line + 1) + ':' + (character + 1));

  return [codeFrame, indentEachLine(location, 1)].join('\n\n');
}

/**
 * @param {Array<TsdResult>} tsdResults
 */
module.exports.formatTsdResults = tsdResults => {
  const title = chalk.bold.red(makeTitle('tsd typecheck'));

  const messages = tsdResults.map(result => {
    const message = ts.flattenDiagnosticMessageText(result.messageText, '\n');

    const codeFrameAndLocation = result.file
      ? getCodeFrameAndLocation(result.file, result.start)
      : undefined;

    if (codeFrameAndLocation !== undefined) {
      return [
        indentEachLine(message, 2),
        indentEachLine(codeFrameAndLocation, 2),
      ].join('\n\n');
    }

    return [indentEachLine(message, 2)].join('\n\n');
  });

  return [title, messages.join('\n\n'), ''].join('\n');
};
