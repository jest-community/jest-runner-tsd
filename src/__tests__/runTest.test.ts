import {
  afterEach,
  beforeEach,
  describe,
  expect,
  jest,
  test,
} from '@jest/globals';
import type { RunTestOptions } from 'create-jest-runner';
import tsdLite, { type TsdResult } from 'tsd-lite';
import type ts from 'typescript';
import { formatTsdResults } from '../formatter';
import runTest from '../run';

jest.mock('tsd-lite');
jest.mock('../formatter', () => ({
  ...jest.requireActual<typeof import('../formatter')>('../formatter'),
  formatTsdResults: jest.fn(() => '<mocked failure message>'),
}));

beforeEach(() => {
  jest.spyOn(Date, 'now').mockReturnValueOnce(1000).mockReturnValueOnce(2800);
});

afterEach(() => {
  jest.clearAllMocks();
});

const runOptions = { config: { slowTestThreshold: 5 } } as RunTestOptions;
const slowRunOptions = { config: { slowTestThreshold: 1 } } as RunTestOptions;

const tsdResults: Array<TsdResult> = [
  {
    file: {} as ts.SourceFile,
    messageText:
      "Argument of type 'string' is not assignable to parameter of type 'number'.",
    start: 64,
  },
  {
    file: {} as ts.SourceFile,
    messageText:
      "Argument of type 'number' is not assignable to parameter of type 'string'.",
    start: 106,
  },
];

describe('runs test', () => {
  test('with all failing results', () => {
    jest
      .mocked(tsdLite)
      .mockReturnValueOnce({ assertionsCount: 2, tsdResults });

    const testResult = runTest({
      ...runOptions,
      testPath: '/path/to/some/failing.test.ts',
    });

    expect(tsdLite).toBeCalledTimes(1);
    expect(tsdLite).toBeCalledWith('/path/to/some/failing.test.ts');

    expect(formatTsdResults).toBeCalledTimes(1);
    expect(formatTsdResults).toBeCalledWith(tsdResults);

    expect(testResult).toMatchSnapshot();
  });

  test('with some failing results', () => {
    jest
      .mocked(tsdLite)
      .mockReturnValueOnce({ assertionsCount: 4, tsdResults });

    const testResult = runTest({
      ...runOptions,
      testPath: '/path/to/some/failing.test.ts',
    });

    expect(tsdLite).toBeCalledTimes(1);
    expect(tsdLite).toBeCalledWith('/path/to/some/failing.test.ts');

    expect(formatTsdResults).toBeCalledTimes(1);
    expect(formatTsdResults).toBeCalledWith(tsdResults);

    expect(testResult).toMatchSnapshot();
  });

  test('with slow results', () => {
    jest
      .mocked(tsdLite)
      .mockReturnValueOnce({ assertionsCount: 2, tsdResults: [] });

    const testResult = runTest({
      ...slowRunOptions,
      testPath: '/path/to/some/slow.test.ts',
    });

    expect(testResult).toMatchObject({
      perfStats: { slow: true },
    });
  });

  test('with all passing results', () => {
    jest
      .mocked(tsdLite)
      .mockReturnValueOnce({ assertionsCount: 2, tsdResults: [] });

    const testResult = runTest({
      ...runOptions,
      testPath: '/path/to/some/passing.test.ts',
    });

    expect(tsdLite).toBeCalledTimes(1);
    expect(tsdLite).toBeCalledWith('/path/to/some/passing.test.ts');

    expect(formatTsdResults).not.toBeCalled();

    expect(testResult).toMatchSnapshot();
  });

  test('with empty results', () => {
    jest
      .mocked(tsdLite)
      .mockReturnValueOnce({ assertionsCount: 0, tsdResults: [] });

    const testResult = runTest({
      ...runOptions,
      testPath: '/path/to/some/empty.test.ts',
    });

    expect(tsdLite).toBeCalledTimes(1);
    expect(tsdLite).toBeCalledWith('/path/to/some/empty.test.ts');

    expect(formatTsdResults).not.toBeCalled();

    expect(testResult).toMatchSnapshot();
  });
});
