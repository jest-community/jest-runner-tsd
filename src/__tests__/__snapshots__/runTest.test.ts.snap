// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`runs test with all failing results 1`] = `
{
  "failureMessage": "<mocked failure message>",
  "leaks": false,
  "numFailingTests": 2,
  "numPassingTests": 0,
  "numPendingTests": 0,
  "numTodoTests": 0,
  "openHandles": [],
  "perfStats": {
    "end": 2800,
    "runtime": 1800,
    "slow": false,
    "start": 1000,
  },
  "skipped": false,
  "snapshot": {
    "added": 0,
    "fileDeleted": false,
    "matched": 0,
    "unchecked": 0,
    "uncheckedKeys": [],
    "unmatched": 0,
    "updated": 0,
  },
  "testFilePath": "/path/to/some/failing.test.ts",
  "testResults": [
    {
      "ancestorTitles": [],
      "duration": 1800,
      "failureDetails": [],
      "failureMessages": [
        "<mocked failure message>",
      ],
      "fullName": "tsd typecheck",
      "numPassingAsserts": 0,
      "status": "failed",
      "title": "tsd typecheck",
    },
  ],
}
`;

exports[`runs test with all passing results 1`] = `
{
  "failureMessage": null,
  "leaks": false,
  "numFailingTests": 0,
  "numPassingTests": 2,
  "numPendingTests": 0,
  "numTodoTests": 0,
  "openHandles": [],
  "perfStats": {
    "end": 2800,
    "runtime": 1800,
    "slow": false,
    "start": 1000,
  },
  "skipped": false,
  "snapshot": {
    "added": 0,
    "fileDeleted": false,
    "matched": 0,
    "unchecked": 0,
    "uncheckedKeys": [],
    "unmatched": 0,
    "updated": 0,
  },
  "testFilePath": "/path/to/some/passing.test.ts",
  "testResults": [
    {
      "ancestorTitles": [],
      "duration": 1800,
      "failureDetails": [],
      "failureMessages": [],
      "fullName": "tsd typecheck",
      "numPassingAsserts": 1,
      "status": "passed",
      "title": "tsd typecheck",
    },
  ],
}
`;

exports[`runs test with empty results 1`] = `
{
  "failureMessage": null,
  "leaks": false,
  "numFailingTests": 0,
  "numPassingTests": 0,
  "numPendingTests": 0,
  "numTodoTests": 0,
  "openHandles": [],
  "perfStats": {
    "end": 2800,
    "runtime": 1800,
    "slow": false,
    "start": 1000,
  },
  "skipped": false,
  "snapshot": {
    "added": 0,
    "fileDeleted": false,
    "matched": 0,
    "unchecked": 0,
    "uncheckedKeys": [],
    "unmatched": 0,
    "updated": 0,
  },
  "testFilePath": "/path/to/some/empty.test.ts",
  "testResults": [],
}
`;

exports[`runs test with some failing results 1`] = `
{
  "failureMessage": "<mocked failure message>",
  "leaks": false,
  "numFailingTests": 2,
  "numPassingTests": 2,
  "numPendingTests": 0,
  "numTodoTests": 0,
  "openHandles": [],
  "perfStats": {
    "end": 2800,
    "runtime": 1800,
    "slow": false,
    "start": 1000,
  },
  "skipped": false,
  "snapshot": {
    "added": 0,
    "fileDeleted": false,
    "matched": 0,
    "unchecked": 0,
    "uncheckedKeys": [],
    "unmatched": 0,
    "updated": 0,
  },
  "testFilePath": "/path/to/some/failing.test.ts",
  "testResults": [
    {
      "ancestorTitles": [],
      "duration": 1800,
      "failureDetails": [],
      "failureMessages": [
        "<mocked failure message>",
      ],
      "fullName": "tsd typecheck",
      "numPassingAsserts": 0,
      "status": "failed",
      "title": "tsd typecheck",
    },
  ],
}
`;
